Proxy identify object until script loads

djih opened the pull request:
Add a lightweight version of identify object to snippet with same interface but no logic. Blocking JS user property operation release.

curtisliu commented on src/amplitude-snippet.js:
```diff
@@ -6,6 +6,11 @@
   as.src = 'https://d24n15hnbwhuhn.cloudfront.net/libs/amplitude-2.4.1-min.gz.js';
   var s = document.getElementsByTagName('script')[0];
   s.parentNode.insertBefore(as, s);
+  amplitude.Identify = function(){this.p={'a':{},'s':{},'u':{},'so':{}};};
```
This is great. I think it would be better to make the list of proxied functions an array in the same manner as line 20 below, and programmatically generate these prototype functions (they all have very similar structure)

djih replied:
Done

djih commented:
Modified snippet to programmatically generate prototype as well as internal property dict for lightweight proxy identify object. The identifyFuncs maps each supported operation to the internal dict key name as well as a boolean which determines whether or not to override the value with '-' (used for the unset operation). This allows for easy addition and removal of operations in the future. 

The same methodology was applied to the Identify class. Added an OPERATIONS dict that maps supported operations to their server-side operation string as well as their corresponding proxy identify dict key (for translation from proxy object). The order of the operations listed in OPERATIONS also determines the order in which they are processed during the translation from proxy object.

Let me know if this is overkill @curtisliu

djih commented on amplitude-segment-snippet.min.js:
```diff
@@ -1,3 +1,6 @@
-(function(e,t){var n=e.amplitude||{};n._q=[];function o(e){n[e]=function(){n._q.push([e].concat(Array.prototype.slice.call(arguments,0)));
-}}var r=["init","logEvent","logRevenue","setUserId","setUserProperties","setOptOut","setVersionName","setDomain","setDeviceId","setGlobalUserProperties"];
-for(var s=0;s<r.length;s++){o(r[s])}e.amplitude=n})(window,document);
+(function(t,e){var n=t.amplitude||{};n.Identify=function(){this.p={};Object.keys(identifyFuncs).forEach(function(t){
```
Somehow I don't think the segment snippet minimize generated properly. I don't see where the identifyFuncs are being declared

djih commented:
Modified identify proxy to use playback queue. PTAL @curtisliu

curtisliu commented:
much cleaner, LGTM.
